kpatch-build/lookup.c

```diff
@@ -407,36 +292,35 @@ void lookup_close(struct lookup_table *table)
 	free(table);
 }
 
-static bool lookup_local_symbol(struct lookup_table *table, char *name,
+static bool lookup_local_symbol(struct lookup_table *table,
+				struct symbol *lookup_sym,
 				struct lookup_result *result)
 {
 	struct object_symbol *sym;
 	unsigned long sympos = 0;
-	int i, in_file = 0;
-
-	if (!table->local_syms)
-		return false;
+	int i;
+	char *file = NULL;
 
 	memset(result, 0, sizeof(*result));
 	for_each_obj_symbol(i, sym, table) {
-		if (sym->bind == STB_LOCAL && !strcmp(sym->name, name))
+		if (sym->bind == STB_LOCAL && !strcmp(sym->name,
+					lookup_sym->name))
 			sympos++;
 
-		if (table->local_syms == sym) {
-			in_file = 1;
+		if (sym->type == STT_FILE) {
+			file = sym->name;
 			continue;
 		}
 
-		if (!in_file)
-			continue;
-
-		if (sym->type == STT_FILE)
-			break;
```

Comment:
I'm not quite convinced this won't cause a regression.  It's possible to have multiple STT_FILE symbols with the same name.  How can we be sure this is the right file?  For example, issue #604.

Reply:
I guessed that shouldn't be that easy but failed to find this case. Pushed v2 that does locals_match() for every STT_FILE section in the file and stores lookup_table->local_syms in each symbol.